### PR TITLE
Add option to template task run name at runtime when using backend API

### LINKED ISSUE #2100: Option to include item string in mapped task name
## Use Case
Minor, non-urgent feature request: mapped tasks are currently referenced in logs and the Cloud UI using their numeric index, e.g. my_task[0], my_task[1]. These numeric values aren't particularly meaningful and, in some cases, the string value of the mapped item would be more useful during log examination or when viewing Flow runs in the Cloud UI.

## Solution
Having an opt-in way to enable use of the string version of a mapped item in the task name would be helpful, e.g.

```
    @task(mapped_item_in_task_name=True)
    def my_task(item):
        ...

    my_task.map(["apple", "banana", "orange"])
```

Which would yield task names of `my_task[apple]`, `my_task[banana]`, and `my_task[orange]` rather than `my_task[0]`, `my_task[1]`. and `my_task[2]`

For tasks mapping over multiple parameters, this option could simply choose to use the first parameter. Elements with long string representations could be truncated to a reasonable size for display purposes. This should *definitely* be opt-in to avoid logging sensitive data.

A better implementation might be to pass an iterable/list to be used for the task names (as a specific kwarg) which might or might not match the parameter to map over (but would need to have the same length):

```
    fruit = {
        0: "apple",
        1: "banana",
        15: "orange"
    }
    my_task.map(fruit.keys(), mapped_task_names=fruit.values())`
```
This would also yield task names of `my_task[apple]`, `my_task[banana]`, and `my_task[orange]`.

## Alternatives
None available currently.

### PR DESCRIPTION
**Pull request by joshmeek**
## Summary
This PR supersedes #2974
Closes #2100 

With the recent introduction of the `set_task_run_name` mutation in the GraphQL API we can now use this to set the name of task runs dynamically at run time. This means that a code snippet like this:

```python
@task
def get_values():
    return ['a', 's', 'd', 'f']

@task(log_stdout=True, task_run_name=lambda **kwargs: f"{kwargs['v']}")
def p(v):
    print(v)

with Flow("task_run_names") as flow:
    vals = get_values()
    p.map(vals)
```

Will automatically name all of the mapped tasks based on their inputs `a`, `s`, `d`, `f` instead of the index!

image

## Changes
Adds a new kwarg called `task_run_name` to the Task class that can now take in a string or callable to update the task run name at runtime.

## Importance
Lots of users have requested this!

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

File: changes/issue2100.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+enhancement:
+  - "Add option to template task run name at runtime when using backend API - [#2100](https://github.com/PrefectHQ/prefect/issues/2100)"
+  - "Add `set_task_run_name` Client function - [#2100](https://github.com/PrefectHQ/prefect/issues/2100)"
+  - "Update mypy typing for `target` kwarg on base Task class - [#2100](https://github.com/PrefectHQ/prefect/issues/2100)"
```
Comment (cicdw) [inline changes/issue2100.yaml:4]: "enhancement"

File: src/prefect/core/task.py [hunk 2]
```diff
@@ -174,7 +175,7 @@ def run(self, x, y):
             is provided, it should have signature `callable(**kwargs) -> str` and at write
             time all formatting kwargs will be passed and a fully formatted location is
             expected as the return value.  Can be used for string formatting logic that
-            `.format(**kwargs)` doesn't support
+            `.format(**kwargs)` supports.
```
Comment (cicdw) [inline src/prefect/core/task.py:178]: I think the intent of this docstring was to say, if you can't do it within `.format()`, you can pass a callable instead

File: src/prefect/core/task.py [hunk 3]
```diff
@@ -188,6 +189,13 @@ def run(self, x, y):
             Task enters a failure state
         - log_stdout (bool, optional): Toggle whether or not to send stdout messages to
             the Prefect logger. Defaults to `False`.
+        - task_run_name (Union[str, Callable], optional): a name to set for this task at runtime.
+            `task_run_name` strings can be templated formatting strings which will be
+            formatted at runtime with values from `prefect.context`. If a callable function
```
Comment (cicdw) [inline src/prefect/core/task.py:194]: also task inputs and parameters!

File: src/prefect/core/task.py [hunk 4]
```diff
@@ -188,6 +189,13 @@ def run(self, x, y):
...
+            is provided, it should have signature `callable(**kwargs) -> str` and at write
+            time all formatting kwargs will be passed and a fully formatted location is
+            expected as the return value. Can be used for string formatting logic that
+            `.format(**kwargs)` supports. Note: this only works for tasks running against a backend API.
```
Comment (cicdw) [inline src/prefect/core/task.py:198]: doesn't support, or we can come up with better wording here

File: src/prefect/core/task.py [hunk 5]
```diff
@@ -216,7 +224,8 @@ def __init__(
         on_failure: Callable = None,
         log_stdout: bool = False,
         result: "Result" = None,
-        target: str = None,
+        target: Union[str, Callable] = None,
```
Comment (cicdw) [inline src/prefect/core/task.py:227]: ENHANCE

**Review by cicdw — CHANGES_REQUESTED** [review]
minor docstring suggestions

File: src/prefect/core/task.py (hunk 2 above)
Comment (joshmeek) [reply src/prefect/core/task.py:178]: Ahhh

**Review by cicdw — APPROVED** [review]
LGTM!  Can't wait to see this in action

**Pull request merged**